### PR TITLE
Enable more warnings and prevent or suppress current ones

### PR DESCRIPTION
**Pull request by Mindavi**
I think it would be good to enable unused variable warnings, to prevent people accidentally sending in patches where they forgot to use a variable.

In some cases (especially with the formatting architecture) it's unavoidable that a format parameter is taken. In those cases I've suppressed the warning with an UNUSED macro.

### CODE REVIEW

**Comment by merbanan** [comment]
I am against void casting in source but wrapping it in an unused macro is ok. One third of the cases can be changed to // comments. The second third the function declaration can be changed and the last third can use the unused macro. I have seen tricks like setting the variable to itself (var=var) to silence these types of warnings. I never liked them because you always get confused about why the code looks the way it does. A macro you can look up and it can have a good comment explaining what it does.

**Comment by Mindavi** [comment]
This is ready for a new review.

File: src/devices/m_bus.c [hunk 1]
```diff
@@ -705,7 +705,7 @@ static int m_bus_mode_s_callback(r_device *decoder, bitbuffer_t *bitbuffer)
 
     // Find a Mode S data package
     unsigned bit_offset = bitbuffer_search(bitbuffer, 0, 0, PREAMBLE_S, sizeof(PREAMBLE_S)*8);
-    start_pos = bitbuffer_manchester_decode(bitbuffer, 0, bit_offset+sizeof(PREAMBLE_S)*8, &packet_bits, 800);
+    bitbuffer_manchester_decode(bitbuffer, 0, bit_offset+sizeof(PREAMBLE_S)*8, &packet_bits, 800);
```
Comment (merbanan) [inline src/devices/m_bus.c:708]: This looks fishy, but I maintain the code so...

**Comment by merbanan** [comment]
Looks acceptable to me. @zuckschwerdt do you have any comments about the changes?

**Comment by zuckschwerdt** [comment]
Tidy cleanup is always nice.
I'm a bit torn about commenting out code (in the decoders) that should work but isn't in the output currently. It pleases the compiler but won't be checked any more thus likely be broken someday.
Some functions should look the same, i.e. dropping `verbose`and `buf_num` isn't great -- the soapy abstraction will likely offer a ring buffer feature soon.
Using `{{0}, 0}` is plain ugly (it assumes too much about the struct), it should be `{0}` but the compiler check for that is broken, `{{0}}` looked like it would work, but I guess doesn't.
The `POSSIBLY_UNUSED(dev);` should likely be a `if (!dev) return -1;`

**Comment by merbanan** [comment]
> I'm a bit torn about commenting out code (in the decoders) that should work but isn't in the output currently. It pleases the compiler but won't be checked any more thus likely be broken someday.

The alternative is to remove it completely or place it under unused macro. It will be in the history to it should be easy to resurrect. But yeah there is no real visually nice solution for this. I do think that more enabled warnings will be better in the long term.

Maybe keep the declaration commented in the decoders is a good enough compromise?

```
    -uint8_t humidity, message_type, l_status;
￼    +uint8_t humidity;

->

    -uint8_t humidity, message_type, l_status;
￼    +// uint8_t message_type, l_status;
￼    +uint8_t humidity;

```

**Comment by Mindavi** [comment]
Keeping the declaration commented is a good enough compromise to me, and helps if someone wants to quickly start working on those parameters in the future (as opposed to searching through git history, which is probably harder in general).

I'll look into the initialization, I'm not too happy about the random brackets and zeroes either. Maybe 'memset'ing to 0 is a better solution? Should typically be very fast to use memset anyway as far as I can tell. Designated initializers are also valid in c99, I read. That could also possibly be a visually more appealing solution.

Yeah, wasn't sure about dropping the parameters for the functions either. I'll re-add them and mark them with an UNUSED macro so the function declarations are consistent.

**Comment by Mindavi** [comment]
Ok, so now memset is used for the initialization of hexstr(s), I've left commented out variables where they were set but unused. I think most comments should be resolved with this.

The only thing I'm not fully sure about is the POSSIBLY_UNUSED macro. @zuckschwerdt thought they might be replaced with `if (!dev) return -1;`, but in some cases a warning message is shown that the function can't be used when certain decoders are not enabled. This leads me to think that the POSSIBLY_UNUSED macro usage is fine. What do you think?

**Comment by zuckschwerdt** [comment]
Sorry about the vague `dev` comment, I only meant the instances in the sdr.c

No way to save the nice `{0}`? Designated initializers is what we use for the r_devices, might be nicer than memset? I.e `= { .p = NULL }` or `= { .idx = 0 }`

**Comment by Mindavi** [comment]
An alternative for the memset would be this:

```diff
diff --git a/src/pulse_analyzer.c b/src/pulse_analyzer.c
index 76bfede..c453b5f 100644
--- a/src/pulse_analyzer.c
+++ b/src/pulse_analyzer.c
@@ -340,8 +340,7 @@ void pulse_analyzer(pulse_data_t *data, int package_type)
     if (hist_timings.bins_count <= 8) {
         // if there is no 3rd gap length output one long B1 code
         if (hist_gaps.bins_count <= 2) {
-            hexstr_t hexstr;
-            memset(&hexstr, 0, sizeof(hexstr));
+            hexstr_t hexstr = {.p = {0}, .idx = 0};
             hexstr_push_byte(&hexstr, 0xaa);
             hexstr_push_byte(&hexstr, 0xb1);
             hexstr_push_byte(&hexstr, hist_timings.bins_count);
@@ -367,8 +366,7 @@ void pulse_analyzer(pulse_data_t *data, int package_type)
             // pick last gap length but a most the 4th
             int limit_bin = MIN(3, hist_gaps.bins_count - 1);
             int limit = hist_gaps.bins[limit_bin].min;
-            hexstr_t hexstrs[HEXSTR_MAX_COUNT];
-            memset(&hexstrs, 0, sizeof(hexstrs));
+            hexstr_t hexstrs[HEXSTR_MAX_COUNT] = {{.p = {0}, .idx = 0}};
             unsigned hexstr_cnt = 0;
             unsigned i = 0;
             while (i < data->num_pulses && hexstr_cnt < HEXSTR_MAX_COUNT) {
```

Does that look better? I'm not sure.

**Comment by zuckschwerdt** [comment]
Do we need to init the whole sub-struct? We just want "default zero everything", so `{0}` is prefered, but due to the compiler mess we fall back to `{{.idx = 0}} // i.e. {0}`.

**Comment by Mindavi** [comment]
Sure can do. I'm not too fond of it, but let's leave it at that syntax (`{.p = {0}}`).

Let's not have too long of a discussion for these details, we can easily change it later if we don't like the current style. I do like the compiler warning when it might not work as expected though, and zero-init is always cleaner that leaving garbage in and getting strange behavior later.

Personally, I think the memset method is the most straightforward. It makes it very explicit the memory is zeroed and doesn't leave anything open to interpretation (is it partially initialized?,...).

Maybe you can take a look at the `dev` usage @zuckschwerdt, I'm not fully sure what to do and the behavior currently isn't changed from what it was before.

Let's go ahead and merge this.

**Comment by merbanan** [comment]
The decoder changes looks good to me.

File: src/data.c [hunk 2]
```diff
@@ -25,6 +25,7 @@
 #include <stdlib.h>
 #include <stdbool.h>
 #include <limits.h>
+
```
Comment (merbanan) [inline src/data.c:28]: Stray change.

Comment (Mindavi) [reply src/data.c:28]: Yeah, that's not really nice

**Pull request merged**